File: Examples/SDKExampleObjC/SDKExampleObjC/Controllers/BindCardViewController.m
//
//  BindCardViewController.m
//  SDKExampleObjC
//
//  Created by Guilherme Rambo on 17/05/21.
//

#import "BindCardViewController.h"

@import Tuna;

@interface BindCardViewController () <UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UITextField *cvvField;
@property (weak, nonatomic) IBOutlet UIButton *bindButton;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;

@end

@implementation BindCardViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.bindButton.enabled = NO;
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (self.card) self.title = self.card.maskedNumber;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    [self.cvvField becomeFirstResponder];
}

- (IBAction)cvvChanged:(UITextField *)sender
{
    self.bindButton.enabled = sender.text.length >= 3;
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [self performBind:textField];
    return YES;
}

- (IBAction)performBind:(id)sender
{
    [self.spinner startAnimating];
    self.bindButton.hidden = YES;
    
    __weak typeof(self) weakSelf = self;
    [TunaSDK bindCard:self.card cvv:self.cvvField.text completion:^(TunaCard *card, TunaSDKError *error) {
        [weakSelf.spinner stopAnimating];
        weakSelf.bindButton.hidden = NO;
        
        if (error) {
            [weakSelf showAlertForError:error];
            return;
        }

        [self.bindButton setTitle:@"Bind Succeeded!" forState:UIControlStateNormal];
        self.bindButton.enabled = NO;
        self.cvvField.enabled = NO;
    }];
}

- (void)showAlertForError:(TunaSDKError *)error
{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Bind Failed" message:[NSString stringWithFormat:@"Bind failed with error: %@", error] preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: Examples/SDKExampleSwift/SDKExampleSwift/BindCardViewController.swift
//
//  BindCardViewController.swift
//  SDKExampleSwift
//
//  Created by Guilherme Rambo on 04/05/21.
//

import UIKit
import Tuna

final class BindCardViewController: UIViewController {
    
    @IBOutlet weak var bindButton: UIButton!
    @IBOutlet weak var spinner: UIActivityIndicatorView!
    @IBOutlet weak var cvvField: UITextField! {
        didSet {
            cvvField?.delegate = self
            cvvField?.addTarget(self, action: #selector(cvvFieldChanged), for: .editingChanged)
        }
    }
    
    var card: TunaCard? {
        didSet {
            guard isViewLoaded else { return }
            
            updateUI()
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        bindButton.isEnabled = false
        updateUI()
    }
    
    private func updateUI() {
        title = card?.maskedNumber ?? "Bind Card"
    }
    
    @IBAction func performBind(_ sender: UIButton) {
        guard let card = card else { return }
        
        spinner.startAnimating()
        sender.isHidden = true
        
        TunaSDK.bindCard(card, cvv: cvvField.text ?? "") { [weak self] result in
            guard let self = self else { return }
            
            defer { self.stopLoading() }
            
            switch result {
            case .success(let bindedCard):
                print(bindedCard)
                self.bindButton.setTitle("Binded Successfully!", for: .disabled)
                self.bindButton.isEnabled = false
            case .failure(let error):
                print(String(describing: error))
                let alert = UIAlertController(title: "Bind Failed", message: String(describing: error), preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                self.present(alert, animated: true, completion: nil)
            }
        }
    }
    
    private func stopLoading() {
        spinner.stopAnimating()
        bindButton.isHidden = false
    }
    
}

extension BindCardViewController: UITextFieldDelegate {
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        performBind(bindButton)
        return true
    }
    
    @objc private func cvvFieldChanged(_ field: UITextField) {
        bindButton.isEnabled = (field.text ?? "").count >= 3
    }
    
}
